Implement size for array namespaces

### Is your feature request related to a problem? Please describe.

When we call `size()` on the torch `array_api_compat` namespace it fails with:

```
if xp.size(ept) > xp.size(p):
       ^^^^^^^
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

self = <earthkit.utils.array.namespace.torch.PatchedTorchNamespace object at 0x105f89400>, name = 'size'

    def __getattr__(self, name):
>       return getattr(self._xp, name)  # Delegate to underlying namespace
               ^^^^^^^^^^^^^^^^^^^^^^^
E       AttributeError: module 'array_api_compat.torch' has no attribute 'size'
```

The reason is that the `size()` is not part of the api standard, but it expects the `size` as an attribute on the array. Howver, for a torch tensor it not an attribute but a method 'my_tensor.size()`!


### Describe the solution you'd like

_No response_

### Describe alternatives you've considered

_No response_

### Additional context

_No response_

### Organisation

ECMWF

--- a/src/earthkit/utils/array/namespace/namespace.py
+++ b/src/earthkit/utils/array/namespace/namespace.py
@@ -80,3 +80,13 @@ class PatchedNamespace:
         Just a placeholder for the numpy function.
         """
         return dict()
+
+    def size(self, x):
+        """Return the size of an array."""
+        x = self._xp.asarray(x)
+        return x.size
+
+    def shape(self, x):
+        """Return the shape of an array."""
+        x = self._xp.asarray(x)
+        return x.shape

--- a/src/earthkit/utils/array/namespace/torch.py
+++ b/src/earthkit/utils/array/namespace/torch.py
@@ -33,3 +33,13 @@ class PatchedTorchNamespace(PatchedNamespace):
         r = self._xp.asarray(r)
         r[self._xp.isnan(x)] = self._xp.nan
         return r
+
+    def size(self, x):
+        """Return the size of an array."""
+        x = self._xp.asarray(x)
+        return x.numel()
+
+    def shape(self, x):
+        """Return the shape of an array."""
+        x = self._xp.asarray(x)
+        return tuple(x.shape)

--- a/tests/test_array.py
+++ b/tests/test_array.py
@@ -158,6 +158,19 @@ def test_patched_namespace_numpy():
 
     assert ns.allclose(ns.pow(c, 2), ns.asarray([1.0, 4.0, 9.0]))
 
+    # size, shape
+    x = ns.asarray([[1.0, 2.0, 3.0], [4.0, 5.0, 6.0]])
+    assert ns.size(x) == 6
+    assert ns.shape(x) == (2, 3)
+
+    x = ns.asarray([3.14])
+    assert ns.size(x) == 1
+    assert ns.shape(x) == (1,)
+
+    x = ns.asarray(3.14)
+    assert ns.size(x) == 1
+    assert ns.shape(x) == tuple()
+
 
 @pytest.mark.skipif(NO_TORCH, reason="No pytorch installed")
 def test_patched_namespace_torch():
@@ -181,6 +194,19 @@ def test_patched_namespace_torch():
     # sign
     x = ns.asarray([1.0, -2.4, ns.nan], dtype=ns.float64)
     assert ns.allclose(ns.sign(x), ns.asarray([1.0, -1.0, ns.nan], dtype=ns.float64), equal_nan=True)
+
+    # size, shape
+    x = ns.asarray([[1.0, 2.0, 3.0], [4.0, 5.0, 6.0]])
+    assert ns.size(x) == 6
+    assert ns.shape(x) == (2, 3)
+
+    x = ns.asarray([3.14])
+    assert ns.size(x) == 1
+    assert ns.shape(x) == (1,)
+
+    x = ns.asarray(3.14)
+    assert ns.size(x) == 1
+    assert ns.shape(x) == tuple()
 
 
 def test_to_device_numpy():

